### DISCUSSION TITLE
Unable to use SSR with Graphcache?

### DISCUSSION THREAD

**Discussion by arogggh** [urql-graphql/urql#3092] [Q&A] [answered]
SSR Exchange doesn't have __typename property on every entity when data from it will be extracted.
After finishing rendering a Vue.js app on the server, I extract all data via the method SSR.extractData() and send it to the client. When the app executes on the client, and I set initialState with data from a server, in the SSR Exchange, I have an error in a console 

> traversal.ts:55 Couldn't find __typename when writing.
> If you're writing to the cache manually have to pass a `__typename` property on each entity in your data.
> (Caused At: "getCategories" query).

I'm using those versions:
@urql/vue: 1.0.5,
@urql/exchange-graphcache: 5.2.0

**Comment by kitten** [comment, marked as answer]
It looks to me like you maybe have the order of exchanges mixed up. Could you post your array of exchanges here please? I think your SSR Exchange may be in front of Graphcache's cacheExchange rather than behind it

**Reply by arogggh** [reply]
You are right, I move Graphcache's cacheExchange before SSR and it's working, thank you so much. But now I have another question.
I'm using Vue 3 with the Suspense feature, and I have duplicated HTTP requests, first on a server and second on a client. Please, could you describe how and what kind of exchanges I should configure, to work it properly?

in a *.vue file, I have a simple lines of code:
`import { useQuery } from @urql/vue`
`const { data, fetching } = await useQuery({
  query: GET_CATEGORIES,
  variables: {
    query: '',
    first: 25,
    after: cursor
  }
})`

On a client, I have extracted data from a server and I didn't expect to execute another http request, when I already have data in a cache.

**Reply by arogggh** [reply]
@kitten, if you need more data or code examples, please let me know. Thank you.

**Reply by kitten** [reply]
You'll have to check whether the data actually ends up in the serialised SSR result and is used properly. I can't really assist you with that effectively just hearing about the "symptoms" 😅 But to give you some pointers,
- the SSR results use the "operation keys" (you'll see `key` properties generated on the query operations) to match the right SSR result
- you have to make sure that the data is serialised after the server-side has completed rendering (i.e. after suspending and when sending the HTML result). Nuxt has a hook for this point in time
- You'll have to check whether `isClient` must be set to `true` on the clientside manually on the `ssrExchange`
- You can use the `debugExchange`, maybe as the first in the chain, to see whether SSR results are used properly and the keys match

**Reply by arogggh** [reply]
Thank you for your response,
Maybe I see what's wrong. I don't see any "operations keys", on SSR result.
I can provide example of SSR result:
`{
  '3801951396': {
    data: '{"categories":{"edges":[{"cursor":"MA==","node":{"id":"2d0d2bb8-1214-4a5e-82f9-5eb43e4a560e","name":"suspense123","slug":"suspense123","published":true,"createdOn":"2023-03-11T10:46:30.051Z","updatedOn":"2023-03-20T16:21:40.929Z","__typename":"CategoryDTO"},"__typename":"CategoriesEdge"},{"cursor":"MQ==","node":{"id":"3f227435-f1c9-4ce1-8ea7-1c944b78f3e3","name":"yesterday-1","slug":"yesterday-1","published":true,"createdOn":"2023-03-07T07:07:18.243Z","updatedOn":"2023-03-07T07:11:14.213Z","__typename":"CategoryDTO"},"__typename":"CategoriesEdge"},{"cursor":"Mg==","node":{"id":"0ad6fd97-825c-40d2-a9c8-12c0c7f4048f","name":"yesterday","slug":"yesterday","published":false,"createdOn":"2023-03-07T07:06:48.022Z","updatedOn":"2023-03-07T07:06:48.022Z","__typename":"CategoryDTO"},"__typename":"CategoriesEdge"}],"pageInfo":{"hasNextPage":true,"endCursor":"Mg==","__typename":"PageInfo"},"totalCount":34,"__typename":"CategoriesConnection"}}'
  }
}`

**Reply by kitten** [reply]
The key here is `3801951396`

**Reply by arogggh** [reply]
> The key here is `3801951396`

That's great, I will continue investigating, where is an issue. Thanks.

**Reply by arogggh** [reply]
I find an issue, it was in my code. During trying to fix Graphcache's cacheExchange with SSR at first, I put clients config: staleWhileRevalidate: true. Now, all good. Thank you so much for helping. Order of exchanges was a root cause.
